README.md
# PSBusinessTime

| Branch | Build status | Last commit | Latest release | PowerShell Gallery | GitHub |
|-|-|-|-|-|-|
| `main` | [![GitHub Workflow Status (main)](https://img.shields.io/github/workflow/status/codaamok/PSBusinessTime/Pipeline/main)](https://github.com/codaamok/PSBusinessTime/actions) | [![GitHub last commit (branch)](https://img.shields.io/github/last-commit/codaamok/PSBusinessTime/main?color=blue)](https://github.com/codaamok/PSBusinessTime/commits/main) | [![GitHub release (latest by date)](https://img.shields.io/github/v/release/codaamok/PSBusinessTime?color=blue)](https://github.com/codaamok/PSBusinessTime/releases/latest) [![GitHub Release Date](https://img.shields.io/github/release-date/codaamok/PSBusinessTime?color=blue)](https://github.com/codaamok/PSBusinessTime/releases/latest) | [![PowerShell Gallery](https://img.shields.io/powershellgallery/dt/PSBusinessTime?color=blue)](https://www.powershellgallery.com/packages/PSBusinessTime) | [![GitHub all releases](https://img.shields.io/github/downloads/codaamok/PSBusinessTime/total?color=blue)](https://github.com/codaamok/PSBusinessTime/releases) |

A PowerShell module with functions to help calculate working hours, and for generally working with or calculating business hours / date times.

Inspired by https://pleasework.robbievance.net/howto-calculate-elapsed-business-hours-using-powershell/.

## Functions

- [Add-WorkingDays](docs/Add-WorkingDays.md)
- [New-BusinessTimeSpan](docs/New-BusinessTimeSpan.md)
- [Get-WorkingDates](docs/Get-WorkingDates.md)
- [Test-WorkingDay](docs/Test-WorkingDay.md)

## Requirements

- PowerShell 5.1 or newer

## Getting started

Install and import:

```powershell
Install-Module PSBusinessTime -Scope CurrentUser
Import-Module PSBusinessTime
```

See the below examples and use `Get-Help` to learn more about the functions with examples. You can discover all functions in the module by running:

```powershell
Get-Command -Module PSBusinessTime
```

## Examples

```powershell
New-BusinessTimeSpan -Start (Get-Date '2022-04-01 00:00:00') -End (Get-Date '2022-04-30 23:59:59') -NonWorkingDates (Get-Date '2022-04-15'), (Get-Date '2022-04-18')
```

The function will return a timespan object of 171 hours:

```powershell
Days              : 7
Hours             : 3
Minutes           : 0
Seconds           : 0
Milliseconds      : 0
Ticks             : 6156000000000
TotalDays         : 7.125
TotalHours        : 171
TotalMinutes      : 10260
TotalSeconds      : 615600
TotalMilliseconds : 615600000
```

2022-04-01 through 2022-04-30 is an entire calendar month, and only 171 hours is considered "working hours" within the defined parameters. '2022-04-15' and '2022-04-18' are considered non-working dates.

___

```powershell
Test-WorkingDay -Date (Get-Date '2022-04-11 09:00:00')
```

The function will return `true` because the datetime is within the default parameters. 2022-04-11 is a Monday, and 09:00 is between 08:00 and 17:00.

___

```powershell
Get-WorkingDates -Start (Get-Date '2022-04-04') -End (Get-Date '2022-04-17') -NonWorkingDaysOfWeek 'Saturday','Sunday','Monday'
```

The function will return an array of 8 datetime objects for '2022-04-05' through to '2022-04-08', and '2022-04-12' through to '2022-04-15':

```powershell
05 April 2022 00:00:00
06 April 2022 00:00:00
07 April 2022 00:00:00
08 April 2022 00:00:00
12 April 2022 00:00:00
13 April 2022 00:00:00
14 April 2022 00:00:00
15 April 2022 00:00:00
```

These are considered working dates within the defined parameters. Saturdays, Sundays, and Mondays, are considered non-working days, therefore every other date inbetween the range is considered a working date.

## To-do

- Use some public API for public holidays
- Allow configurable 'lunch hour' and so it respects the designated time period, too, e.g.
  - if start date is 8am and end date is 5pm, and 'lunch hour' time period is 12pm - 1pm, then working hours is 8.
  - if start date is 12pm and end date is 5pm, and 'lunch hour' time period is 12pm - 1pm, then working hours is 4.
- Allow config for specifying days with different working hours, e.g. perhaps specific dates of the month or year might be 4 or 5 working hours rather than 8 or 9, which is generally days like Christmas Eve is most western countries


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.2] - 2022-10-05
### Fixed
- `New-BusinessTimeSpan` threw an exception in some scenarios for Windows PowerShell - see issue #1

## [0.3.0] - 2022-05-04
### Changed
- Renamed `Get-ElapsedBusinessTime` to `New-BusinessTimeSpan`, and changed parameters `-StartDate` and `-EndDate` to `-Start` and `-End`, similar to `New-TimeSpan` and its datetime parameters.

## [0.2.2] - 2022-04-15
### Fixed
- CmdletBinding attribute was in incorrect position for `Add-WorkingDays`. This caused PlatyPS to not generate docs properly. This fix does not provide any functional improvement to `Add-WorkingDays`. At this time, it is just necessary to push to source control which invokes the release pipeline.

## [0.2.1] - 2022-04-15
### Fixed
- `New-BusinessTimeSpan` did not return correct result if `-StartDate` and `-EndDate` were on the same day, but it was a working day, and its time for both dates were outside of working hours

## [0.2.0] - 2022-04-14
### Added
- New function `Add-WorkingDays`

## [0.1.7] - 2022-04-14
### Changed
- `Test-WorkingDay` now has parameter validation to ensure `-StartHour` is less then `-FinishHour`
- `New-BusinessTimeSpan` now has parameter validation to ensure `-StartHour` is less then `-FinishHour`

## [0.1.6] - 2022-04-14
### Fixed
- `New-BusinessTimeSpan` did not return correct result if `-StartDate` and `-EndDate` were on the same day, but it was a working day, and its time for both dates were outside of working hours

## [0.1.5] - 2022-04-14
### Fixed
- `New-BusinessTimeSpan` did not return correct result if range between `-StartHour` and `-FinishHour` was not whole, ie was a span of time where minutes and seconds also needed consideration

## [0.1.4] - 2022-04-14
### Fixed
- `New-BusinessTimeSpan` did not return correct result when either the start date or end date were outside of business hours
- Added line breaks in between each line for all comment based help

## [0.1.2] - 2022-04-13
### Fixed
- `New-BusinessTimeSpan` did not return correct result where the number of working days between `-StartDate` and `-EndDate` was 1 but the date range spanned more than 1 calendar day, e.g. Sunday through to Monday.

## [0.1.1] - 2022-04-12
### Fixed
- Some typo and grammar issues in comment based help
- `Get-WorkingDates` would not return the correct number of working dates, which also impacted `New-BusinessTimeSpan`

## [0.1.0] - 2022-04-11
### Added
- Initial release

[Unreleased]: https://github.com/codaamok/PSBusinessTime/compare/0.3.2..HEAD
[0.3.2]: https://github.com/codaamok/PSBusinessTime/compare/0.3.0..0.3.2
[0.3.0]: https://github.com/codaamok/PSBusinessTime/compare/0.2.2..0.3.0
[0.2.2]: https://github.com/codaamok/PSBusinessTime/compare/0.2.1..0.2.2
[0.2.1]: https://github.com/codaamok/PSBusinessTime/compare/0.2.0..0.2.1
[0.2.0]: https://github.com/codaamok/PSBusinessTime/compare/0.1.7..0.2.0
[0.1.7]: https://github.com/codaamok/PSBusinessTime/compare/0.1.6..0.1.7
[0.1.6]: https://github.com/codaamok/PSBusinessTime/compare/0.1.5..0.1.6
[0.1.5]: https://github.com/codaamok/PSBusinessTime/compare/0.1.4..0.1.5
[0.1.4]: https://github.com/codaamok/PSBusinessTime/compare/0.1.2..0.1.4
[0.1.2]: https://github.com/codaamok/PSBusinessTime/compare/0.1.1..0.1.2
[0.1.1]: https://github.com/codaamok/PSBusinessTime/compare/0.1.0..0.1.1
[0.1.0]: https://github.com/codaamok/PSBusinessTime/tree/0.1.0

tests/invoke.tests.ps1
$PesterConfig = New-PesterConfiguration -Hashtable @{
    Run = @{
        Path = '{0}/Public' -f $PSScriptRoot
        Throw = $true
        #SkipRemainingOnFailure = 'Block'
    }
    Output = @{
        Verbosity = 'Detailed'
    }
    TestResult = @{
        Enabled = $true
        OutputFormat = 'NUnit2.5'
    }
}

Invoke-Pester -Configuration $PesterConfig -ErrorAction 'Stop'

tests/Public/Add-WorkingDays.Tests.ps1
BeforeAll {
    # Always use built code if running in a pipeline
    if ($env:USER -eq 'runner') {
        Import-Module "$PSScriptRoot/../../build/PSBusinessTime/PSBusinessTime.psd1" -Force
    }
    # Check if module is already imported, as it can be via VSCode task where you can choose what code base to test
    # and you might not want to cloober it with the non-built code
    elseif (-not (Get-Module PSBusinessTime)) {
        Import-Module "$PSScriptRoot/../../src/PSBusinessTime.psd1" -Force
    }
}

Describe "Add-WorkingDays" {
    It "should be 2022-04-19 as 3 working days from 2022-04-14" {
        Add-WorkingDays -Date (Get-Date '2022-04-14') -Days 3 | Should -Be (Get-Date '2022-04-19')
    }

    It "should be 2022-04-11 as -3 working days from 2022-04-14" {
        Add-WorkingDays -Date (Get-Date '2022-04-14') -Days -3 | Should -Be (Get-Date '2022-04-11')
    }

    It "should be 2022-04-21 as 3 working days from 2022-04-14, with Good Friday and Easter Monday holidays" {
        Add-WorkingDays -Date (Get-Date '2022-04-14') -Days 3 -NonWorkingDates (Get-Date '2022-04-15'),(Get-Date '2022-04-18') | Should -Be (Get-Date '2022-04-21')
    }

    It "should be 2022-04-21 as 3 working days from 2022-04-14, excluding Friday, Saturday, Sunday, and Monday, as working days" {
        Add-WorkingDays -Date (Get-Date '2022-04-14') -Days 3 -NonWorkingDaysOfWeek 'Friday','Saturday','Sunday','Monday' | Should -Be (Get-Date '2022-04-21')
    }
}

tests/Public/Get-WorkingDays.Tests.ps1
BeforeAll {
    # Always use built code if running in a pipeline
    if ($env:USER -eq 'runner') {
        Import-Module "$PSScriptRoot/../../build/PSBusinessTime/PSBusinessTime.psd1" -Force
    }
    # Check if module is already imported, as it can be via VSCode task where you can choose what code base to test
    # and you might not want to cloober it with the non-built code
    elseif (-not (Get-Module PSBusinessTime)) {
        Import-Module "$PSScriptRoot/../../src/PSBusinessTime.psd1" -Force
    }
}

Describe "Get-WorkingDays" {
    It "should be 0 working days" {
        $StartDate = Get-Date '2022-04-09 08:00:00'
        $EndDate   = Get-Date '2022-04-09 17:00:00'
        Get-WorkingDays -StartDate $StartDate -EndDate $EndDate | Should -Be 0
    }

    It "should be 1 working day" {
        $StartDate = Get-Date '2022-04-07 08:00:00'
        $EndDate   = Get-Date '2022-04-07 17:00:00'
        Get-WorkingDays -StartDate $StartDate -EndDate $EndDate | Should -Be 1
    }

    It "should be 2 working days across 2 consecutive days" {
        $StartDate = Get-Date '2022-04-07 08:00:00'
        $EndDate   = Get-Date '2022-04-08 17:00:00'
        Get-WorkingDays -StartDate $StartDate -EndDate $EndDate | Should -Be 2
    }

    It "should be 2 working days across 3 consecutive days, with traditional weekends" {
        $StartDate = Get-Date '2022-04-07 08:00:00'
        $EndDate   = Get-Date '2022-04-09 17:00:00'
        Get-WorkingDays -StartDate $StartDate -EndDate $EndDate | Should -Be 2
    }

    It "should be 10 working days across 14 consecutive days, with traditional weekends" {
        $StartDate = Get-Date '2022-04-04 08:00:00'
        $EndDate   = Get-Date '2022-04-17 17:00:00'
        Get-WorkingDays -StartDate $StartDate -EndDate $EndDate | Should -Be 10
    }

    It "should be 9 working days across 14 consecutive days, where a date inbetween is a non-working date eg public holiday" {
        $StartDate = Get-Date '2022-04-04 08:00:00'
        $EndDate   = Get-Date '2022-04-17 17:00:00'
        $NonWorkingDates = Get-Date '2022-04-05'
        Get-WorkingDays -StartDate $StartDate -EndDate $EndDate -NonWorkingDates $NonWorkingDates | Should -Be 9
    }

    It "should be 8 working days across 14 consecutive days, where Sat+Sun+Mon are non-working days" {
        $StartDate = Get-Date '2022-04-04 08:00:00'
        $EndDate   = Get-Date '2022-04-17 17:00:00'
        Get-WorkingDays -StartDate $StartDate -EndDate $EndDate -NonWorkingDaysOfWeek 0,1,6 | Should -Be 8
    }
}

src/Public/Add-WorkingDays.ps1
function Add-WorkingDays {
    <#
    .SYNOPSIS
        Add a number of working days onto a given date.
    .DESCRIPTION
        Add a number of working days onto a given date.

        What constitutes a "working day" in terms of day of the week, or calendar date, is arbitrary and completely customisable.

        In other words, the default parameters dictate normal working days, which are Monday through Friday.

        You can also specify particular dates, or days of the week, to be regarded as non-working dates via the -NonWorkingDates and -NonWorkingDaysOfWeek parameters.
    .PARAMETER Date
        The starting date used for calculation.
        
        The default value is the current datetime.
    .PARAMETER Days
        The number of working days to add onto from the given date.
        
        Number can be negative in order to substract from the given date, too.
    .PARAMETER NonWorkingDaysOfWeek
        The days of the week, representated as strings e.g. 'Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday', which denotes non-working days of the week.
        
        Days specified in this parameter will not be considered as working days.

        Default values are Saturday and Sunday.
    .PARAMETER NonWorkingDates
        An array of datetime objects which denote specific non-working dates.
        
        Dates specified in this parameter will not be considered as working days.
    .EXAMPLE
        Add-WorkingDays -Days 3
        
        Adds 3 working days onto the current date. For example, if today's date is 2022-04-07, then the returned datetime object will be 2022-04-12.
    .EXAMPLE
        Add-WorkingDays -Days -3

        Minuses 3 working days from the current date. For example, if today's date is 2022-04-07, then the returned datetime object will be 2022-04-04.
    .EXAMPLE
        Add-WorkingDays -Date (Get-Date '2022-04-14') -Days 5 -NonWorkingDates (Get-Date '2022-04-15'),(Get-Date '2022-04-18')

        Add 5 working days from 2022-04-14, discounting 2022-04-15 (Good Friday) and 2022-04-18 (Easter Monday) as working days. The returned datetime object will be 2022-04-25.
    .EXAMPLE
        Add-WorkingDays -Days 1 -NonWorkingDaysOfWeek 'Friday','Saturday','Sunday'

        Add 1 working day onto the current date. For example, if today's date is 2022-04-07, then the returned datetime object will be 2022-04-11.
    .INPUTS
        This function does not accept pipeline input.
    .OUTPUTS
        System.DateTime
    .NOTES
        Chris Dent (@indented-automation) wrote this in the WinAdmins Discord
        https://discord.com/channels/618712310185197588/618857671608500234/913855890384371712
    #>
    [CmdletBinding()]
    param (
        [Parameter()]
        [object]$Date = (Get-Date),

        [Parameter(Mandatory)]
        [int]$Days,

        [Parameter()]
        [ValidateSet('Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday')]
        [String[]]$NonWorkingDaysOfWeek = @('Saturday','Sunday'),

        [Parameter()]
        [DateTime[]]$NonWorkingDates
    )

    $increment = $Days / [Math]::Abs($Days)
    do {
        $Date = $Date.AddDays($increment)
        
        if ($NonWorkingDaysOfWeek -notcontains $Date.DayOfWeek -And $NonWorkingDates -notcontains $Date.Date) {
            $Days -= $increment
        }
    } while ($Days)

    return $Date
}


src/Public/Get-WorkingDays.ps1
function Get-WorkingDays {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory)]
        [DateTime]$StartDate,

        [Parameter(Mandatory)]
        [ValidateScript({
            if ($StartDate -gt $_) { throw "-StartDate must be less than -EndDate." } else { return $true }
        })]
        [DateTime]$EndDate,

        [Parameter()]
        [int[]]$NonWorkingDaysOfWeek = @(0,6),

        [Parameter()]
        [DateTime[]]$NonWorkingDates = @()
    )
 
    if ($StartDate.TimeOfDay -eq $EndDate.TimeOfDay) { 
        # This can return 1 less than intended if we do not do this change. 
        # For example, if the dates in between are 3 working days,
        # but the time span between them are 2 whole days,
        # this returns 2 instead of 3
        $EndDate = $EndDate.AddSeconds(1) 
    }

    $TimeSpan = New-TimeSpan -Start $StartDate -End $EndDate
    $Days = [Math]::Ceiling($TimeSpan.TotalDays)
    $Date = $StartDate

    $WorkingDays = do {
        if ($NonWorkingDaysOfWeek -notcontains $Date.DayOfWeek -And $NonWorkingDates -notcontains $Date.Date) {
            $Date
        }
        $Date = $Date.AddDays(1)
        $Days--
    } while ($Days)

    $WorkingDays.Count
}

src/Public/Test-WorkingDay.ps1
function Test-WorkingDay {
    <#
    .SYNOPSIS
        Determine whether a given datetime is a working day.
    .DESCRIPTION
        Determine whether a given datetime is a working day.
        
        What constitutes a "working day" in terms of day of the week, or calendar date, including working hours, is arbitrary and completely customisable.
        
        In other words, the default parameters dictate normal working days, which are Monday through Friday, and normal working hours are 08:00 through 17:00.
        
        You can also specify particular dates, or days of the week, to be regarded as non-working dates via the -NonWorkingDates and -NonWorkingDaysOfWeek parameters.
        
        If the datetime of -Date falls outside of these parameters, you'll receive a boolean result.
    .PARAMETER Date
        The datetime to determine whether it is a working day or not.
    .PARAMETER StartHour
        The starting hour of a typical working day. The default starting hour is 08:00 (AM).
        
        Note: this parameter is a datetime object is, however only the time is used for calculation. The date is ignored.
    .PARAMETER FinishHour
        The final hour of a typical working day. The default final hour is 17:00.
        
        Note: this parameter is a datetime object is, however only the time is used for calculation. The date is ignored.
    .PARAMETER NonWorkingDaysOfWeek
        The days of the week, representated as strings e.g. 'Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday', which denotes non-working days of the week.
        
        Days specified in this parameter will not be considered as working days.

        Default values are Saturday and Sunday.
    .PARAMETER NonWorkingDates
        An array of datetime objects which denote specific non-working dates.
        
        Dates specified in this parameter will not be considered as working days.
    .EXAMPLE
        Test-WorkingDay -Date (Get-Date '2022-04-11 09:00:00')
        
        The function will return true because the datetime is within the default parameters. 2022-04-11 is a Monday, and 09:00 is between 08:00 and 17:00.
    .EXAMPLE
        Test-WorkingDay -Date (Get-Date '2022-04-10 11:00:00')

        The function will return false because the datetime is outside the default parameters. 2022-04-10 is a Sunday, and therefore is not a working day, regardless of the hour/minute.
    .EXAMPLE
        Test-WorkingDay -Date (Get-Date '2022-04-11 22:00:00') -StartHour (Get-Date '08:00:00') -FinishHour Get-Date '23:00:00'

        The function will return true because the datetime is within the defined parameters. 2022-04-11 is a Monday, and 22:00 is between 08:00 and 23:00.
        Note: a datetime object is passed for both -StartHour and -FinishHour, however only the time is used for calculation. The date is ignored.
    .EXAMPLE
        Test-WorkingDay -Date (Get-Date '2022-04-11 09:00:00') -NonWorkingDaysOfWeek 'Saturday','Sunday','Monday'

        The function will return false because the datetime is outside the defined parameters. 2022-04-11 is a Monday, and is considered a non working day of the week.
    .EXAMPLE
        Test-WorkingDay -Date (Get-Date '2022-04-11 09:00:00') -NonWorkingDates (Get-Date '2022-04-11')

        The function will return false because the datetime is outside the defined parameters. 2022-04-11 is a Monday, and is considered a non working date.
    .INPUTS
        This function does not accept pipeline input.
    .OUTPUTS
        System.Boolean
    #>
    [CmdletBinding()]
    param (
        [Parameter(Mandatory)]
        [DateTime]$Date,

        [Parameter()]
        [DateTime]$StartHour = '08:00:00',

        [Parameter()]
        [ValidateScript({
            if ($StartHour -gt $_) { throw "-StartHour must be less than -FinishHour." } else { return $true }
        })]
        [DateTime]$FinishHour = '17:00:00',

        [Parameter()]
        [ValidateSet('Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday')]
        [String[]]$NonWorkingDaysOfWeek = @('Saturday','Sunday'),

        [Parameter()]
        [DateTime[]]$NonWorkingDates
    )

    $NonWorkingDaysOfWeek -notcontains $Date.DayOfWeek -And
    $Date.TimeOfDay -ge $StartHour.TimeOfDay -And
    $Date.TimeOfDay -lt $FinishHour.TimeOfDay -And
    $NonWorkingDates -notcontains $Date.Date
}

src/Public/Get-ElapsedBusinessTime.ps1
function Get-ElapsedBusinessTime {
    <#
    .SYNOPSIS
        Get the elapsed time between two dates, where the time measured is only inbetween "business hours".
    .DESCRIPTION
        Get the elapsed time between two dates, where the time measured is only inbetween "business hours".

        This is helpful to measure the amount of time past from a start datetime, to an end datetime, while only considering "business hours".

        What constitutes "business hours" in terms of day of the week, or calendar date, including working hours, is arbitrary and completely customisable.

        In other words, the default parameters dictate normal working days, which are Monday through Friday and 08:00 through 17:00.

        You can also specify particular dates, or days of the week, to be regarded as non-working dates via the -NonWorkingDates and -NonWorkingDaysOfWeek parameters.

        This function does consider both date and time while calculating the elapsed time.
    .PARAMETER StartDate
        The datetime object to start calculating the elapsed time from. It must be an older datetime than -EndDate.
    .PARAMETER EndDate
        The datetime object to end calculating the elapsed time to. It must be a newer datetime than -StartDate.
    .PARAMETER StartHour
        The starting hour of a typical working day. The default starting hour is 08:00 (AM).
        
        Note: this parameter is a datetime object is, however only the time is used for calculation. The date is ignored.
    .PARAMETER FinishHour
        The final hour of a typical working day. The default final hour is 17:00.
                
        Note: this parameter is a datetime object is, however only the time is used for calculation. The date is ignored.
    .PARAMETER NonWorkingDaysOfWeek
        The days of the week, representated as strings e.g. 'Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday', which denotes non-working days of the week.
        
        Days specified in this parameter will not be considered as working days.

        Default values are Saturday and Sunday.
    .PARAMETER NonWorkingDates
        An array of datetime objects which denote specific non-working dates.
        
        Dates specified in this parameter will not be considered as working days.
    .EXAMPLE
        Get-ElapsedBusinessTime -StartDate (Get-Date '2022-04-11 10:00:00') -EndDate (Get-Date '2022-04-11 10:37:00')
        
        The function will return a timespan object of 37 minutes. 2022-04-11 is a Monday and the whole time inbetween the date range given is within the default parameters.
    .EXAMPLE
        Get-ElapsedBusinessTime -StartDate (Get-Date '2022-04-11 08:00:00') -EndDate (Get-Date '2022-04-12 08:00:00')
        
        The function will return a timespan object of 9 hours. 2022-04-11 is a Monday and 2022-04-12 is a Tuesday, and only 9 hours is considered "working hours" within the default parameters. 
    .EXAMPLE
        Get-ElapsedBusinessTime -StartDate (Get-Date '2022-04-11 13:00:00') -EndDate (Get-Date '2022-04-13 13:00:00')
        
        The function will return a timespan object of 18 hours. 2022-04-11 through 2022-04-13 is Monday through Wednesday, and only 18 hours is considered "working hours" within the default parameters. 
    .EXAMPLE
        Get-ElapsedBusinessTime -StartDate (Get-Date '2022-04-01 00:00:00') -EndDate (Get-Date '2022-04-30 23:59:59') -NonWorkingDates (Get-Date '2022-04-15'), (Get-Date '2022-04-18')
        
        The function will return a timespan object of 162 hours. 2022-04-01 through 2022-04-30 is an entire calendar month, and only 162 hours is considered "working hours" within the defined parameters. '2022-04-15' and '2022-04-18' are considered non-working dates.
    .EXAMPLE
        Get-ElapsedBusinessTime -StartDate (Get-Date '2022-01-01 00:00:00') -EndDate (Get-Date '2022-12-31 23:59:59') -NonWorkingDates (Get-Date '2022-01-03'), (Get-Date '2022-04-15'), (Get-Date '2022-04-18'), (Get-Date '2022-05-02'), (Get-Date '2022-06-02'), (Get-Date '2022-06-03'), (Get-Date '2022-08-29'), (Get-Date '2022-12-26'), (Get-Date '2022-12-27')
        
        The function will return a timespan object of 2259 hours. 2022-01-01 through 2022-12-31 is an entire year, and only 2259 hours is considered "working hours" within the defined parameters. All dates passed to -NonWorkingDates are considered non-working dates (public holidays in the UK for 2022).
    .EXAMPLE
        Get-ElapsedBusinessTime -StartDate (Get-Date '2022-01-01 00:00:00') -EndDate (Get-Date '2022-12-31 23:59:59') -NonWorkingDaysOfWeek 'Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday'
        
        The function will return a timespan object of 0 hours. 2022-01-01 through 2022-12-31 is an entire year, and 0 hours is considered "working hours" within the defined parameters. All days passed to -NonWorkingDaysOfWeek are considered non-working days, hence the result of 0 hours.
    .INPUTS
        This function does not accept pipeline input.
    .OUTPUTS
        System.TimeSpan
    #>
    [CmdletBinding()]
    param(
        [Parameter(Mandatory)]
        [datetime]$StartDate,
        
        [Parameter(Mandatory)]
        [ValidateScript({
            if ($StartDate -gt $_) { throw "-StartDate must be less than -EndDate." } else { return $true }
        })]
        [datetime]$EndDate,

        [Parameter()]
        [DateTime]$StartHour = '08:00:00',

        [Parameter()]
        [ValidateScript({
            if ($StartHour -gt $_) { throw "-StartHour must be less than -FinishHour." } else { return $true }
        })]
        [DateTime]$FinishHour = '17:00:00',

        [Parameter()]
        [ValidateSet('Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday')]
        [String[]]$NonWorkingDaysOfWeek = @('Saturday','Sunday'),

        [Parameter()]
        [DateTime[]]$NonWorkingDates
    )

    $CommonParams = @{
        NonWorkingDaysOfWeek = $NonWorkingDaysOfWeek
        NonWorkingDates      = $NonWorkingDates
    }

    $WorkingHours = New-TimeSpan -Start $StartHour -End $FinishHour
    $WorkingDays  = Get-WorkingDates -StartDate $StartDate -EndDate $EndDate @CommonParams

    if ($null -eq $WorkingDays) {
        New-TimeSpan
    }
    elseif ($WorkingDays.Count -eq 1) {
        $Params = @{
            StartHour = $StartHour
            FinishHour = $FinishHour
        }

        $_StartDate = Get-Date ('{0}/{1}/{2} {3}:{4}:{5}' -f $WorkingDays.Year,
                                                                $WorkingDays.Month,
                                                                $WorkingDays.Day,
                                                                $StartHour.Hour,
                                                                $StartHour.Minute,
                                                                $StartHour.Second)

        if ($StartDate -le $_StartDate) {
            $Params["StartDate"] = $_StartDate
        }
        else {
            $Params["StartDate"] = $StartDate
        }

        $_EndDate = Get-Date ('{0}/{1}/{2} {3}:{4}:{5}' -f $WorkingDays.Year,
                                                              $WorkingDays.Month,
                                                              $WorkingDays.Day,
                                                              $FinishHour.Hour,
                                                              $FinishHour.Minute,
                                                              $FinishHour.Second)

        if ($EndDate -gt $_EndDate) {
            $Params["EndDate"] = $_EndDate
        }
        else {
            $Params["EndDate"] = $EndDate
        }

        $Result = GetElapsedTime @Params

        if ($Result.Ticks -le 0) {
            New-TimeSpan
        }
        else {
            $Result
        }
    }
    else {
        $NumberOfWorkingDays = $WorkingDays.Count
        $ElapsedTime = New-TimeSpan
        $InBetweenHours = New-TimeSpan
        
        $FirstDayEndDate = Get-Date ('{0}/{1}/{2} {3}:{4}:{5}' -f $StartDate.Year,
                                                                  $StartDate.Month,
                                                                  $StartDate.Day,
                                                                  $FinishHour.Hour, 
                                                                  $FinishHour.Minute, 
                                                                  $FinishHour.Second)

        if (Test-WorkingDay -Date $StartDate -StartHour $StartHour -FinishHour $FinishHour @CommonParams) {        
            $Params = @{
                StartDate  = $StartDate
                EndDate    = $FirstDayEndDate
                StartHour  = $StartHour
                FinishHour = $FinishHour
            }
            $ElapsedTime += (GetElapsedTime @Params)
            $NumberOfWorkingDays--
        }
        elseif ($StartDate -gt $FirstDayEndDate) {
            $NumberOfWorkingDays--
        }

        $LastDayStartDate = Get-Date ('{0}/{1}/{2} {3}:{4}:{5}' -f $EndDate.Year,
                                                                       $EndDate.Month,
                                                                       $EndDate.Day,
                                                                       $StartHour.Hour, 
                                                                       $StartHour.Minute, 
                                                                       $StartHour.Second)

        if (Test-WorkingDay -Date $EndDate -StartHour $StartHour -FinishHour $FinishHour @CommonParams) {
            $Params = @{
                StartDate  = $LastDayStartDate
                EndDate    = $EndDate
                StartHour  = $StartHour
                FinishHour = $FinishHour
            }
            $ElapsedTime += (GetElapsedTime @Params)
            $NumberOfWorkingDays--
        }
        elseif ($EndDate -lt $LastDayStartDate) {
            $NumberOfWorkingDays--
        }

        $InBetweenHours = $NumberOfWorkingDays * $WorkingHours
        $InBetweenHours + $ElapsedTime
    }
}